command/agent/config/config.go

```diff
@@ -169,14 +171,253 @@ func NewConfig() *Config {
 	}
 }
 
+// Merge merges two Agent configurations.
+func (c *Config) Merge(c2 *Config) *Config {
+	if c2 == nil {
+		return c
+	}
+
+	result := NewConfig()
+
+	result.SharedConfig = c.SharedConfig
+	if c2.SharedConfig != nil {
+		result.SharedConfig = c.SharedConfig.Merge(c2.SharedConfig)
+	}
+
+	result.AutoAuth = c.AutoAuth
+	if c2.AutoAuth != nil {
+		result.AutoAuth = c2.AutoAuth
+	}
+
+	result.Cache = c.Cache
+	if c2.Cache != nil {
+		result.Cache = c2.Cache
+	}
+
+	result.APIProxy = c.APIProxy
+	if c2.APIProxy != nil {
+		result.APIProxy = c2.APIProxy
+	}
+
+	result.DisableMlock = c.DisableMlock
+	if c2.DisableMlock {
+		result.DisableMlock = c2.DisableMlock
+	}
+
+	// For these, ignore the non-specific one and overwrite them all
+	result.DisableIdleConnsAutoAuth = c.DisableIdleConnsAutoAuth
+	if c2.DisableIdleConnsAutoAuth {
+		result.DisableIdleConnsAutoAuth = c2.DisableIdleConnsAutoAuth
+	}
+
+	result.DisableIdleConnsAPIProxy = c.DisableIdleConnsAPIProxy
+	if c2.DisableIdleConnsAPIProxy {
+		result.DisableIdleConnsAPIProxy = c2.DisableIdleConnsAPIProxy
+	}
+
+	result.DisableIdleConnsTemplating = c.DisableIdleConnsTemplating
+	if c2.DisableIdleConnsTemplating {
+		result.DisableIdleConnsTemplating = c2.DisableIdleConnsTemplating
+	}
+
+	result.DisableKeepAlivesAutoAuth = c.DisableKeepAlivesAutoAuth
+	if c2.DisableKeepAlivesAutoAuth {
+		result.DisableKeepAlivesAutoAuth = c2.DisableKeepAlivesAutoAuth
+	}
+
+	result.DisableKeepAlivesAPIProxy = c.DisableKeepAlivesAPIProxy
+	if c2.DisableKeepAlivesAPIProxy {
+		result.DisableKeepAlivesAPIProxy = c2.DisableKeepAlivesAPIProxy
+	}
+
+	result.DisableKeepAlivesTemplating = c.DisableKeepAlivesTemplating
+	if c2.DisableKeepAlivesTemplating {
+		result.DisableKeepAlivesTemplating = c2.DisableKeepAlivesTemplating
+	}
+
+	result.TemplateConfig = c.TemplateConfig
+	if c2.TemplateConfig != nil {
+		result.TemplateConfig = c2.TemplateConfig
+	}
+
+	for _, l := range c.Templates {
+		result.Templates = append(result.Templates, l)
+	}
+	for _, l := range c2.Templates {
+		result.Templates = append(result.Templates, l)
+	}
+
+	result.ExitAfterAuth = c.ExitAfterAuth
+	if c2.ExitAfterAuth {
+		result.ExitAfterAuth = c2.ExitAfterAuth
+	}
+
+	result.Vault = c.Vault
+	if c2.Vault != nil {
+		result.Vault = c2.Vault
+	}
+
+	result.PidFile = c.PidFile
+	if c2.PidFile != "" {
+		result.PidFile = c2.PidFile
+	}
+
+	return result
+}
+
+// ValidateConfig validates an Agent configuration after it has been fully merged together, to
+// ensure that required combinations of configs are there
+func (c *Config) ValidateConfig() error {
+	if c.APIProxy != nil && c.Cache != nil {
+		if c.Cache.UseAutoAuthTokenRaw != nil {
+			if c.APIProxy.UseAutoAuthTokenRaw != nil {
+				return fmt.Errorf("use_auto_auth_token defined in both api_proxy and cache config. Please remove this configuration from the cache block")
+			} else {
+				c.APIProxy.ForceAutoAuthToken = c.Cache.ForceAutoAuthToken
```

Comment:
I'd prefer if a Validate method didn't mutate its inputs.

Reply:
This is an unfortunate consequence of supporting the API Proxy config values in Cache, too. It made the most sense to me to keep this logic together as opposed to having two separate blocks e.g.:
```
// in Validate
	if c.APIProxy != nil && c.Cache != nil {
		if c.Cache.UseAutoAuthTokenRaw != nil {
			if c.APIProxy.UseAutoAuthTokenRaw != nil {
				return fmt.Errorf("use_auto_auth_token defined in both api_proxy and cache config. Please remove this configuration from the cache block")
			}
		}
	}
```
and
```
// elsewhere
	if c.APIProxy != nil && c.Cache != nil {
		if c.Cache.UseAutoAuthTokenRaw != nil {
			if c.APIProxy.UseAutoAuthTokenRaw == nil {
				c.APIProxy.ForceAutoAuthToken = c.Cache.ForceAutoAuthToken
			}
		}
	}
```

I don't really like mutation like this either, but in this case it feels like the lesser of two evils to me, with regards to code cleanliness.